engine/jobmaster/dm/dm_jobmaster.go

```diff
@@ -331,3 +342,64 @@ func (jm *JobMaster) preCheck(ctx context.Context, cfg *config.JobCfg) error {
 	jm.Logger().Info("finish pre-checking job config", zap.String("result", msg))
 	return nil
 }
+
+// all task finished and all worker tombstone
+func (jm *JobMaster) isFinished(ctx context.Context) bool {
+	return jm.taskManager.allFinished(ctx) && jm.workerManager.allTombStone()
+}
+
+func (jm *JobMaster) status(ctx context.Context, code frameModel.WorkerStatusCode) (frameModel.WorkerStatus, error) {
+	status := frameModel.WorkerStatus{
+		Code: code,
+	}
+	if jobStatus, err := jm.QueryJobStatus(ctx, nil); err != nil {
+		return status, err
+	} else if bs, err := json.Marshal(jobStatus); err != nil {
+		return status, err
+	} else {
+		status.ExtBytes = bs
+		return status, nil
+	}
+}
+
+// cancel remove jobCfg in metadata, and wait all workers offline.
+func (jm *JobMaster) cancel(ctx context.Context, code frameModel.WorkerStatusCode) error {
```

Comment:
Is `cancel` reentrant? When Tick and OnCancel concurrently call it

Reply:
yes